zeroize/tests/zeroize_derive.rs

```diff
@@ -285,4 +285,49 @@ mod custom_derive_tests {
         }
         assert_eq!(&value.0 .0, &[0, 0, 0])
     }
+
+    #[test]
+    fn check_actual_memory_vec() {
+        #[derive(Zeroize, ZeroizeOnDrop)]
+        struct X([u8; 3]);
+
+        #[derive(Zeroize, ZeroizeOnDrop)]
+        struct Z(X);
+
+        let mut value = Z(X([1, 2, 3]));
+        let actual_memory_content = unsafe {
+            let b = value.0 .0;
+            std::slice::from_raw_parts(b.as_ptr(), b.len())
+        };
+
+        assert_eq!(actual_memory_content, value.0 .0);
+        assert_eq!(&actual_memory_content, &[1, 2, 3]);
+
+        drop(value);
+
+        assert_eq!(&actual_memory_content, &[0, 0, 0])
```

Comment:
This is a use-after-free and there are no guarantees the value will actually be zero. You can do this instead:

```suggestion
        assert_ne!(&actual_memory_content, &[1, 2, 3])
```

You might try using a larger array to reduce the possibility that it's overwritten with something that might just so happen to be the expected pattern to be cleared.

Reply:
I thought, maybe because I reused `let mut value = Z(X([1, 2, 3]));` and it's just coincidence? But switching to `[4, 5, 8]` (`git grep` says this is unique across the crate) yields same result:

```
failures:

---- custom_derive_tests::check_actual_memory_vec stdout ----
thread 'custom_derive_tests::check_actual_memory_vec' panicked at 'assertion failed: `(left != right)`
  left: `[4, 5, 8]`,
 right: `[4, 5, 8]`', zeroize/tests/zeroize_derive.rs:308:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    custom_derive_tests::check_actual_memory_vec

test result: FAILED. 14 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

error: test failed, to rerun pass '--test zeroize_derive'
```

Reply:
The problem I'm trying to highlight here is the use-after-free.

You are storing a pointer to potentially dropped memory and trying to read its value.

All bets are off after that. It can be arbitrarily overwritten.

My point is your 3-byte sequence is small enough it's possible that it just-so-happens it is overwritten with that value. I'm not saying that's actually what's happening here: see my explanation below. But to write a reliable test you need to lower the probability that it isn't being overwritten by a value which just so happens to match the original pattern.

Reply:
Yes, I follow you, but even with increasingly unlikely coincidences I still see the same failure, which is why I am inclined to think something else is happening and drew my comparison about `.0 .0` and `.0`. Re "see my explanation below" I am not seeing alternative explanation comment on this page. I'll check Conversation tab.

```
---- custom_derive_tests::check_actual_memory_vec stdout ----
thread 'custom_derive_tests::check_actual_memory_vec' panicked at 'assertion failed: `(left != right)`
  left: `[4, 5, 8, 23, 44, 101, 44, 0, 39]`,
 right: `[4, 5, 8, 23, 44, 101, 44, 0, 39]`', zeroize/tests/zeroize_derive.rs:308:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
Try replacing `drop` with a scope and see what happens.